quarkchain/cluster/master.py

```diff
@@ -107,119 +108,127 @@ def __init__(self, header, peer):
             Logger.log_exception()
             self.peer.close_with_error(str(e))
 
+    async def __download_block_header_and_check(self, start, skip, limit):
+        _, resp, _ = await self.peer.write_rpc_request(
+            op=CommandOp.GET_ROOT_BLOCK_HEADER_LIST_WITH_SKIP_REQUEST,
+            cmd=GetRootBlockHeaderListWithSkipRequest.create_for_height(
+                height=start,
+                skip=skip,
+                limit=limit,
+                direction=Direction.TIP
+            )
+        )
+
+        if resp.root_tip.total_difficulty < self.header.total_difficulty:
+            raise RuntimeError("Bad peer sending root block tip with lower TD")
+
+        # new limit should equal to limit, but in case that remote has chain reorg,
+        # the remote tip may has lower height and greater TD.
+        new_limit = min(limit, len(range(start, resp.root_tip.height + 1, skip + 1)))
+        if len(resp.block_header_list) != new_limit:
+            # Something bad happens
+            raise RuntimeError("Bad peer sending incorrect number of root block headers")
+
+        return resp
+
+    async def __find_ancestor(self):
+        # n-ary search
+        start = max(self.root_state.tip.height - self.root_state.root_config.MAX_STALE_ROOT_BLOCK_HEIGHT_DIFF, 0)
+        end = min(self.root_state.tip.height, self.header.height)
+
+        while end >= start:
+            span = (end - start) // ROOT_BLOCK_HEADER_LIST_LIMIT + 1
+            resp = await self.__download_block_header_and_check(start, span - 1, len(range(start, end + 1, span)))
+
+            if len(resp.block_header_list) == 0:
```

Comment:
nit: `len(range(start, end + 1, span))` can probably just be `ROOT_BLOCK_HEADER_LIST_LIMIT`

Reply:
end  - start may be smaller than ROOT_BLOCK_HEADER_LIST_LIMIT (e.g., end == start) so we should not ask for maximum headers in each request.

Reply:
ah, right